### PR TITLE
Replace Eigen with inhouse vector and matrix classes

### PR DESCRIPTION
**Pull request by tomaz82**
This patch removes Eigen as a dependency. Eigen itself is a quite large library with a lot of optimized math functions, which ES uses none at all. The few things it does use has no benefits of the optimizations that Eigen can do. These 5 classes replace the few pieces that ES actually use at a much smaller footprint and way simpler (and potentially faster) code. Where Eigen is a maze of templates and inheritance this is simple and straightforward.

As an added bonus this also drops the compile time by about 30%.

### CODE REVIEW

**Comment by zigurana** [comment]
Impressive! The reduction in compile time is certainly welcome!

I've quickly scanned your additions (the new `math` methods), and it seems very rigorously done. I am no expert though, so I'd like some pointers. How do you suggest we could try and test this (big) change?

**Comment by tomaz82** [comment]
The vector and matrix math is mostly used by GUI objects, so what I did was to test several different themes and made sure they looked identical before and after this change.

**Comment by joolswills** [comment]
Thanks for this.

**Comment by jrassa** [comment]
I see that Vector4 has a v3 method that casts from Vector4 to Vector3.  It would be great if Vector3 had a similar v2 method.  This is something I could never figure out how to do with Eigen.

It would also be nice if there was a way to go the other way, from Vector2 to Vector3.
```cpp
// Vector2 -> Vector3 

// replace this
Vector3f vec3(vec2.x(), vec2.y(), 0.0);
// with
Vector 3f vec3(v2) // z defaults to 0.0
// or
Vector3f vec3(v2, 0.0)
// or
Vector3f v3 = v2.v3(0.0);
```

None of this is a blocker and I could always add it on my own later, but I wanted to throw it out there.

I will try to test this out soon.

File: es-core/src/math/Matrix4x4.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+		mZ.v3() *= scale;
+		mW.v3() *= scale;
+
+		return *this;
+	}
+
+	M& rotate(const S angle, const Vector3<S>& axis)
+	{
+		const S pi = (S)3.1415926535897932384626433832795028841971693993751058209749445923;
+		const S	a  = -angle * (pi / 180);
```
Comment (jrassa) [inline es-core/src/math/Matrix4x4.h]: I believe Eigen's rotate method expected angles to already be in radians and as such that is what we are already passing in.

**Comment by hex007** [comment]
@tomaz82 Once you are done with code commits squash the commits so it can be merged.

**Comment by joolswills** [comment]
@hex007 It doesn't need to be squashed if all the commits are organised as this looks. Only needed when multiple commits are fixes for earlier ones etc. Although there is a merge commit that could be removed.

**Comment by tomaz82** [comment]
Just want to let you all know that I'm currently looking into all that has been mentioned, just gonna make sure it all works as intended and then I'll update the PR.

File: es-core/src/math/Matrix4x4.h (hunk 1 above)
Comment (hex007) [reply es-core/src/math/Matrix4x4.h]: Should Pi be moved to a static class member considering that it will never change?

**Comment by hex007** [comment]
@tomaz82 Just a reminder that CMakeList.txt still marks Eigen3 as a dependency. You will need to take that out, delete your build files and run `cmake ..` again to clear build cache.

**Comment by tomaz82** [comment]
@hex007 The PI stuff is gone since ES already has the angles as radians. I don't use CMake tho, I handcrafted a MSVC project. But yeah, I'll cleanup the CMakeList.txt to remove the dependency on Eigen3.

**Comment by tomaz82** [comment]
Made a new PR because I goofed up with git
https://github.com/RetroPie/EmulationStation/pull/263

**Comment by joolswills** [comment]
You can force push back to this branch from a fixed local branch. No need for new pr.

**Comment by tomaz82** [comment]
@joolswills Yeah, that was my initial problem, but the reason i made a new PR is because I somehow managed to delete this branch from my github.

**Comment by joolswills** [comment]
Can be restored - https://stackoverflow.com/questions/17929891/for-how-long-can-you-restore-recover-a-deleted-branch-on-github

**Pull request closed without merging**